File: Leetcode/longest_common_prefix.py
# https://leetcode.com/problems/longest-common-prefix/
'''
    Author - Subhajit Das
    University of Engineering and Management, Kolkata
    01/09/2019
'''


class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        prefix = ""
        if(len(strs) == 0):
            return prefix
        elif(len(strs) == 1):
            return strs[0]

        prev = ""
        idx = 0
        while(True):
            for word in strs:
                if(len(word)-1 >= idx):
                    if(prev == ""):
                        prev = word[idx]
                    elif(prev != word[idx]):
                        return prefix
                else:
                    return prefix
            prefix += strs[0][idx]
            prev = ""
            idx += 1

File: InterviewBit/LongestCommonPrefix.java
// https://www.interviewbit.com/problems/longest-common-prefix/
/*
    Author @Subhajit Das (sdsubhajitdas.github.io)
    SWE @Turbot HQ India PVT Ltd.
    27/03/2021
*/

import java.util.*;

public class Solution {
    public String longestCommonPrefix(ArrayList<String> words) {
        boolean matched = true;
        String prefix = "";
        int idx = 0, minLength = Integer.MAX_VALUE;

        for(String word: words)
            minLength = Math.min(minLength, word.length());

        while(matched && idx < minLength) {
            String newPrefix = prefix + words.get(0).charAt(idx);

            for(String word: words)
                if(!word.startsWith(newPrefix)){
                    matched = false;
                    break;
                }
            if(matched)
                prefix = newPrefix;
            idx++;
        }
        return prefix;
    }
}
